File: simulate.py
import random
from collections import Counter

DIE_SIDES = [4, 6, 8, 10, 10, 12, 20]
D20 = 6
WIN_GOALS = 3
Z_IDS = ["-3", "-2", "-1", "1", "2", "3"]

def roll(s): return random.randint(1, s)
def make_dice(): return [{"id": i, "sides": s, "value": 0, "rest_until": 0} for i, s in enumerate(DIE_SIDES)]
def is_avail(d, pos): return pos > d["rest_until"]
def use_die(d, pos): d["rest_until"] = pos + 1

def get_next_zone(curr, dir_):
    if curr == 0: return "1" if dir_ == 1 else "-1"
    idx = Z_IDS.index(str(curr))
    nxt = idx + dir_
    if nxt < 0 or nxt >= len(Z_IDS): return None
    return Z_IDS[nxt]

def simulate_game():
    h_dice = make_dice()
    v_dice = make_dice()
    scores = {"h": 0, "v": 0}
    turn = 0
    possession = 1
    ball_pos = 0  # persists across possessions now

    while scores["h"] < WIN_GOALS and scores["v"] < WIN_GOALS:
        pos = possession
        wall = {z: 0 for z in Z_IDS}

        # Re-roll dice returning from rest
        for dice in (h_dice, v_dice):
            for d in dice:
                if d["rest_until"] == pos - 1:
                    d["value"] = roll(d["sides"])

        scored = False
        turnover = False

        if turn == 0:
            # Home (human) attacks — Visitor (AI) defends
            # AI only defends zones attacker still needs to breach
            d20v = v_dice[D20]
            if is_avail(d20v, pos): wall["3"] = d20v["value"]; use_die(d20v, pos)

            avail_v = sorted([d for d in v_dice if is_avail(d, pos)], key=lambda d: -d["value"])
            i = 0
            if ball_pos < 2:  # zone 2 not yet bypassed
                if len(avail_v) > i: wall["2"] += avail_v[i]["value"]; use_die(avail_v[i], pos); i += 1
                if len(avail_v) > i: wall["2"] += avail_v[i]["value"]; use_die(avail_v[i], pos); i += 1
            if ball_pos < 1:  # zone 1 not yet bypassed
                if len(avail_v) > i: wall["1"] = avail_v[i]["value"]; use_die(avail_v[i], pos); i += 1

            # Human attacks greedily from current ball_pos
            bp = ball_pos
            while not scored and not turnover:
                target = get_next_zone(bp, 1)
                if target is None: turnover = True; break
                wall_val = wall[target]
                avail_h = sorted([d for d in h_dice if is_avail(d, pos)], key=lambda d: -d["value"])
                if not avail_h: turnover = True; break
                total, used = 0, []
                for d in avail_h:
                    if len(used) >= 3: break
                    total += d["value"]; used.append(d)
                    if total > wall_val: break
                for d in used: use_die(d, pos)
                if total > wall_val:
                    bp = int(target)
                    if bp == 3: scores["h"] += 1; scored = True
                else:
                    turnover = True

            ball_pos = 0 if scored else bp

        else:
            # Visitor (AI) attacks — Home (human) defends
            # Human mirrors AI: only defend negative zones not yet bypassed
            d20h = h_dice[D20]
            if is_avail(d20h, pos): wall["-3"] = d20h["value"]; use_die(d20h, pos)

            avail_h = sorted([d for d in h_dice if is_avail(d, pos)], key=lambda d: -d["value"])
            i = 0
            if ball_pos >= -1:  # zone -2 not yet bypassed
                if len(avail_h) > i: wall["-2"] += avail_h[i]["value"]; use_die(avail_h[i], pos); i += 1
                if len(avail_h) > i: wall["-2"] += avail_h[i]["value"]; use_die(avail_h[i], pos); i += 1
            if ball_pos >= 0:  # zone -1 not yet bypassed
                if len(avail_h) > i: wall["-1"] = avail_h[i]["value"]; use_die(avail_h[i], pos); i += 1

            # AI (visitor) attacks greedily from current ball_pos
            bp = ball_pos
            while not scored and not turnover:
                target = get_next_zone(bp, -1)
                if target is None: turnover = True; break
                wall_val = wall[target]
                avail_v = sorted([d for d in v_dice if is_avail(d, pos)], key=lambda d: -d["value"])
                if not avail_v: turnover = True; break
                total, used = 0, []
                for d in avail_v:
                    if len(used) >= 3: break
                    total += d["value"]; used.append(d)
                    if total > wall_val: break
                for d in used: use_die(d, pos)
                if total > wall_val:
                    bp = int(target)
                    if bp == -3: scores["v"] += 1; scored = True
                else:
                    turnover = True

            ball_pos = 0 if scored else bp

        if scores["h"] >= WIN_GOALS or scores["v"] >= WIN_GOALS:
            break
        possession += 1
        turn = 1 - turn

    return possession, scores["h"], scores["v"]


random.seed(42)
N = 10_000
results = [simulate_game() for _ in range(N)]
poss_list = [r[0] for r in results]
poss_list_sorted = sorted(poss_list)

home_wins   = sum(1 for _, h, v in results if h > v)
visitor_wins = sum(1 for _, h, v in results if v > h)
score_counts = Counter((h, v) for _, h, v in results)

avg = sum(poss_list) / N
p50 = poss_list_sorted[N // 2]
p90 = poss_list_sorted[int(N * 0.90)]
p95 = poss_list_sorted[int(N * 0.95)]
p99 = poss_list_sorted[int(N * 0.99)]

print(f"\n=== 10,000 games · ball stays on turnover ===")
print(f"Avg possessions : {avg:.1f}")
print(f"Median          : {p50}")
print(f"P90 / P95 / P99 : {p90} / {p95} / {p99}")
print(f"Min / Max       : {poss_list_sorted[0]} / {poss_list_sorted[-1]}")
print(f"\nHome wins    : {home_wins:,} ({home_wins/N*100:.1f}%)")
print(f"Visitor wins : {visitor_wins:,} ({visitor_wins/N*100:.1f}%)")
print(f"\nScore breakdown:")
for score, cnt in sorted(score_counts.items(), key=lambda x: -x[1])[:15]:
    bar = '█' * (cnt // 60)
    print(f"  {score[0]}-{score[1]} : {cnt:>5} ({cnt/N*100:.1f}%)  {bar}")

File: simulate.js
// Park the Bus — 10,000-game simulation
// Human (Home) uses greedy strategy mirroring the AI

const DIE_SIDES = [4, 6, 8, 10, 10, 12, 20];
const D20 = 6;
const WIN_GOALS = 3;
const Z_IDS = ["-3", "-2", "-1", "1", "2", "3"];

const rollDie = s => Math.floor(Math.random() * s) + 1;
const makeDice = () => DIE_SIDES.map((s, i) => ({ id: i, sides: s, value: 0, restUntil: 0 }));
const isAvail = (d, pos) => pos > d.restUntil;
const useDie = (d, pos) => { d.restUntil = pos + 1; };

function getNextZone(curr, dir) {
    if (curr === 0) return dir === 1 ? "1" : "-1";
    const idx = Z_IDS.indexOf(curr.toString());
    const next = idx + dir;
    if (next < 0 || next >= Z_IDS.length) return null;
    return Z_IDS[next];
}

function simulateGame() {
    const state = {
        scores: { h: 0, v: 0 },
        turn: 0,
        hDice: makeDice(),
        vDice: makeDice(),
        possession: 1,
    };

    while (state.scores.h < WIN_GOALS && state.scores.v < WIN_GOALS) {
        const pos = state.possession;
        const wall = { "-3": 0, "-2": 0, "-1": 0, "1": 0, "2": 0, "3": 0 };

        // Re-roll dice returning from rest (restUntil === pos-1), skip unused (restUntil < pos-1)
        [state.hDice, state.vDice].forEach(dice =>
            dice.forEach(d => {
                if (d.restUntil === pos - 1) d.value = rollDie(d.sides);
            })
        );

        let scored = false;
        let turnover = false;

        if (state.turn === 0) {
            // Home attacks (human) — Visitor defends (AI sets wall)
            const d20v = state.vDice[D20];
            if (isAvail(d20v, pos)) { wall["3"] = d20v.value; useDie(d20v, pos); }

            const availV = state.vDice.filter(d => isAvail(d, pos)).sort((a, b) => b.value - a.value);
            if (availV[0]) { wall["2"] += availV[0].value; useDie(availV[0], pos); }
            if (availV[1]) { wall["2"] += availV[1].value; useDie(availV[1], pos); }
            if (availV[2]) { wall["1"]  = availV[2].value; useDie(availV[2], pos); }

            // Human attacks: greedy, up to 3 dice, must use at least 1
            let ballPos = 0;
            while (!scored && !turnover) {
                const target = getNextZone(ballPos, 1);
                if (!target) { turnover = true; break; }

                const wallVal = wall[target] || 0;
                const availH = state.hDice.filter(d => isAvail(d, pos)).sort((a, b) => b.value - a.value);
                if (availH.length === 0) { turnover = true; break; }

                let sum = 0, used = [];
                for (const d of availH) {
                    if (used.length >= 3) break;
                    sum += d.value; used.push(d);
                    if (sum > wallVal) break;
                }
                used.forEach(d => useDie(d, pos));

                if (sum > wallVal) {
                    ballPos = parseInt(target);
                    if (ballPos === 3) { state.scores.h++; scored = true; }
                } else {
                    turnover = true;
                }
            }

        } else {
            // Visitor attacks (AI) — Home defends (human mirrors AI strategy)
            const d20h = state.hDice[D20];
            if (isAvail(d20h, pos)) { wall["-3"] = d20h.value; useDie(d20h, pos); }

            const availH = state.hDice.filter(d => isAvail(d, pos)).sort((a, b) => b.value - a.value);
            if (availH[0]) { wall["-2"] += availH[0].value; useDie(availH[0], pos); }
            if (availH[1]) { wall["-2"] += availH[1].value; useDie(availH[1], pos); }
            if (availH[2]) { wall["-1"]  = availH[2].value; useDie(availH[2], pos); }

            // AI attacks: greedy, up to 3 dice
            let ballPos = 0;
            while (!scored && !turnover) {
                const target = getNextZone(ballPos, -1);
                if (!target) { turnover = true; break; }

                const wallVal = wall[target] || 0;
                const availV = state.vDice.filter(d => isAvail(d, pos)).sort((a, b) => b.value - a.value);
                if (availV.length === 0) { turnover = true; break; }

                let sum = 0, used = [];
                for (const d of availV) {
                    if (used.length >= 3) break;
                    sum += d.value; used.push(d);
                    if (sum > wallVal) break;
                }
                used.forEach(d => useDie(d, pos));

                if (sum > wallVal) {
                    ballPos = parseInt(target);
                    if (ballPos === -3) { state.scores.v++; scored = true; }
                } else {
                    turnover = true;
                }
            }
        }

        if (state.scores.h >= WIN_GOALS || state.scores.v >= WIN_GOALS) break;
        state.possession++;
        state.turn = 1 - state.turn;
    }

    return { possessions: state.possession, h: state.scores.h, v: state.scores.v };
}

const N = 10000;
let totalPoss = 0, homeWins = 0, visitorWins = 0;
const possHist = {};

for (let i = 0; i < N; i++) {
    const r = simulateGame();
    totalPoss += r.possessions;
    if (r.h > r.v) homeWins++;
    else visitorWins++;
    possHist[r.possessions] = (possHist[r.possessions] || 0) + 1;
}

const avg = totalPoss / N;
const sorted = Object.keys(possHist).map(Number).sort((a, b) => a - b);
let cumul = 0, p50, p95;
for (const k of sorted) {
    cumul += possHist[k];
    if (!p50 && cumul >= N * 0.5) p50 = k;
    if (!p95 && cumul >= N * 0.95) p95 = k;
}

console.log(`\n=== Park the Bus — ${N.toLocaleString()} game simulation ===`);
console.log(`Average possessions per game : ${avg.toFixed(2)}`);
console.log(`Median possessions           : ${p50}`);
console.log(`95th-percentile possessions  : ${p95}`);
console.log(`Min / Max                    : ${sorted[0]} / ${sorted[sorted.length - 1]}`);
console.log(`\nHome (human) wins  : ${homeWins.toLocaleString()} (${(homeWins/N*100).toFixed(1)}%)`);
console.log(`Visitor (AI) wins  : ${visitorWins.toLocaleString()} (${(visitorWins/N*100).toFixed(1)}%)`);
